lib/spack/spack/build_environment.py

```diff
@@ -474,6 +470,41 @@ def static_to_shared_library(static_lib, shared_lib=None, **kwargs):
     m.static_to_shared_library = static_to_shared_library
 
 
+def get_parallel_make_jobs(pkg):
+    """Returns the number of jobs to be used to build this package
+
+    Args:
+        pkg: package to be built
+    """
+    # If pkg.parallel is False, the package can't be built in parallel
+    if not pkg.parallel:
+        return 1
+
+    # Get the maximum number of build jobs, as defined in the config file
+    max_build_jobs = spack.config.get('config:max_build_jobs') or \
+        multiprocessing.cpu_count()
+
+    # The default number of build jobs is the minimum between max_build_jobs
+    # and the available number of cores
+    default_build_jobs = min(max_build_jobs, multiprocessing.cpu_count())
+
+    # The requested number of build jobs is either pkg.make_jobs (usually
+    # a cli request) or 'config:build_jobs' or the default
+    build_jobs = pkg.make_jobs or spack.config.get('config:build_jobs') \
+        or default_build_jobs
```

Comment:
There's a lot of complexity that has accumulated here.  `pkg.make_jobs` probably shouldn't be the way this is set from the command line anymore now that we have command line scopes.  I think the command-line preference for `make_jobs` should come in through the config system like other command-line overrides.

Look at how `debug`, `locks`, and `verify_ssl` are handled in `main.py`.  I think `make_jobs` should be set via:

```python
 spack.config.set('config:make_jobs', args.jobs, scope='command_line')
```

And `-j` / `--jobs` should probably become a common argument so that other commands (like `diy`, etc.) can use it.  That `config.set` can probably be part of a custom argument action in `spack.cmd.common.arguments`.

`build_environment._set_variables_for_single_module()` should *set* `make_jobs` in the build environment at module scope, but we can get rid of it as a package attribute.  Packages should never set it -- it doesn't make sense, and AFAICT none currently do.  It should *only* come from config (files/envs/command line).

Does that make sense?  It would let us get rid of some complexity in the packages, and it would remove at least one thing that we currently have to pass to the package in the install command.

Reply:
> Does that make sense? It would get rid of some complexity in the packages, and it would remove at least one thing that we currently have to pass through in the install command.

This all makes sense to me. Being a refactoring that is not tied in any way to the issue being fixed, would you prefer having a separate PR for it?

Reply:
@alalazo: up to you whether you want to do it separately or not; I think it’s probably easier to do it all at once.

@healther: the scope argument is for blame reporting so that `spack config blame` can tell the user how the option was set.  I’m not sure there’s a good what to set it automatically... there isn’t an “active” scope.

Reply:
Waiting for #11524 to be merged to move this PR forward